Classification/include/CGAL/Classification/TensorFlow/Neural_network_classifier.h

```diff
@@ -0,0 +1,1122 @@
+  {
+    clear();
+  }
+
+  bool initialized() const { return (m_root != NULL); }
+
+  bool& verbose() { return m_verbose; }
+
+  void clear()
+  {
+    clear (m_ph_gt);
+    clear (m_ph_ft);
+    clear (m_loss);
+    clear (m_session);
+    clear (m_root);
+    m_weights.clear();
+    m_bias.clear();
+    m_optimizer.clear();
+    m_layers.clear();
+  }
+
+  template <typename T>
+  void clear (T* t)
+  {
+    if (t != NULL)
+      delete t;
+    t = NULL;
+  }
+
+  void compute_normalization_coefficients (const std::vector<std::size_t>& indices)
+  {
+//#define DO_NOT_NORMALIZE_FEATURES
+#ifdef DO_NOT_NORMALIZE_FEATURES
+    m_feature_means.clear();
+    m_feature_means.resize (m_features.size(), 0.f);
+    m_feature_sd.clear();
+    m_feature_sd.resize (m_features.size(), 1.f);
+#else
+    m_feature_means.clear();
+    m_feature_means.resize (m_features.size(), 0.f);
+    
+    for (std::size_t i = 0; i < indices.size(); ++ i)
+      for (std::size_t f = 0; f < m_features.size(); ++ f)
+        m_feature_means[f] += m_features[f]->value(indices[i]);
+    for (std::size_t f = 0; f < m_features.size(); ++ f)
+      m_feature_means[f] /= indices.size();
+
+    m_feature_sd.clear();
+    m_feature_sd.resize (m_features.size(), 0.f);
+
+    for (std::size_t i = 0; i < indices.size(); ++ i)
+      for (std::size_t f = 0; f < m_features.size(); ++ f)
+        m_feature_sd[f] +=
+          (m_features[f]->value(indices[i]) - m_feature_means[f]) *
+          (m_features[f]->value(indices[i]) - m_feature_means[f]);
+    for (std::size_t f = 0; f < m_features.size(); ++ f)
+    {
+      m_feature_sd[f] = std::sqrt (m_feature_sd[f] / indices.size());
+      if (m_feature_sd[f] == 0.f)
+        m_feature_sd[f] = 1.f;
+
+      if (std::isnan(m_feature_means[f]))
+        m_feature_means[f] = 0.f;
+      if (std::isnan(m_feature_sd[f]))
+        m_feature_sd[f] = 1.f;
+
+      // if (m_verbose)
+      //   std::cerr << "#" << f << ": " << m_features[f]->name() << " = "
+      //             << m_feature_means[f] << " +/- " << m_feature_sd[f] << std::endl;
+    }
+#endif
+  }
+  /// \endcond
+  
+  /// @}
+
+  /// \name Training
+  
+  /// @{
+  /*!
+    \brief Runs the training algorithm.
+
+    From the set of provided ground truth, this algorithm constructs a
+    neural network and applies an Adam optimizer to set up the weights
+    and bias that produce the most accurate result with respect to
+    this ground truth.
+
+    \pre At least one ground truth item should be assigned to each
+    label.
+
+    \param ground_truth vector of label indices. It should contain for
+    each input item, in the same order as the input set, the index of
+    the corresponding label in the `Label_set` provided in the
+    constructor. Input items that do not have a ground truth
+    information should be given the value `-1`.
+
+    \param restart_from_scratch should be set to `false` if the user
+    wants to continue adjusting weights and bias based, and kept to
```

Comment:
set vs kept, set in both cases?

Reply:
"Kept" was to indicate that it was the default value, but okay.